docs: add deployment and technical documentation (#47)
rdest-devs/cogniro: docs/DEPLOYMENT.md

   stores all data as files in one directory on disk. There is no database to install.
2. **The frontend**: a React website built with Next.js. After it is built it is just static files
   (HTML, CSS, JavaScript, images). It listens on port **3000** in the simple setup, or it is
   served by your web server. It needs no runtime once built.

Participants and the administrator use a web browser. The browser loads the frontend, and the
frontend calls the backend over HTTP.

```
  Browser  ->  Frontend (static files)  ->  Backend API (port 8000)  ->  Data directory on disk
```

You will typically put both behind one HTTPS reverse proxy (for example Nginx) so that
everything is served from your domain over HTTPS. Section 8 shows how.

---

## 2. Prerequisites

### 2.1 For Option A (Docker)

- A Linux host (a VM or server) with:
  - Docker Engine (recent version).
  - Docker Compose v2 (the `docker compose` command).
- Outbound internet access during the build (to download base images and packages).

That is all. The runtimes (Python, Node) are inside the images.

### 2.2 For Option B (without Docker)

On the host that runs the **backend**:

> - Python **3.14** (this exact minor version is required).

Reviewer: Unneccessary
```suggestion
- Python **3.14**
```

- `uv`, the Python package manager from Astral (it creates the virtual environment and

>   installs dependencies). Install it from https://docs.astral.sh/uv/ or with

Reviewer: Everybody knows what is a package manager
```suggestion
- `uv`, the Python package manager from Astral. Install it from https://docs.astral.sh/uv/ or with
```

  `curl -LsSf https://astral.sh/uv/install.sh | sh`.

On the machine that **builds** the frontend (can be the same host, or a CI runner, or your
laptop):

- Node.js **22**.
- `pnpm` **10.33.0**. The easiest way is `corepack enable && corepack prepare pnpm@10.33.0 --activate`.

On the host that **serves** the frontend in production:

- A web server that can serve static files (for example Nginx, Apache, Caddy, or a CDN). No
  Node.js is required to serve the built site. If you only want a quick start without a web
  server, you can serve the static files with the `serve` package (via `pnpm dlx`) instead
  (section 7.3).

---

## 3. Get the code

Clone the repository onto the host (or build machine):

```bash

> git clone <your-repo-url> cogniro

Reviewer: ?

cd cogniro
```

The two parts live in `backend/` and `frontend/`. Documentation (including this file) is in
`docs/`.

---

## 4. Create the required secrets

The backend needs two secrets before it will serve admin login in production:

1. `ADMIN_PASSWORD_HASH` - a bcrypt hash of the admin password (never the plaintext).
2. `JWT_SECRET` - a long random string used to sign login tokens.

### 4.1 Generate the admin password hash

The repository ships a helper script. Run it from the `backend` directory. It will prompt you
for the password and print the hash.

With Docker available (no local Python needed):

```bash
docker run --rm -it -v "$PWD/backend:/app" -w /app python:3.14-slim \
  sh -c "pip install --quiet bcrypt && python scripts/hash_admin_password.py"
```

Or, on a host that has Python and uv (Option B):

```bash
cd backend
uv run python scripts/hash_admin_password.py

> ```

Reviewer: uv init && uv sync?

Copy the printed hash. It looks like `$2b$12$....`. You will paste it into
`ADMIN_PASSWORD_HASH`. Always keep it **single-quoted** in env files and compose files so the
shell does not try to expand the `$` characters.

### 4.2 Generate the JWT secret

Any long random value works. For example:

```bash
openssl rand -base64 48
```

Copy the output into `JWT_SECRET`.

Keep both secrets out of version control. Do not commit a real `.env` file (the repository's
`.gitignore` already excludes `.env`).

---

## 5. Environment variables

> This section lists what you must set. The backend reads its configuration from environment
> variables; a full annotated template is in `backend/.env.example`. The frontend reads a few
> `NEXT_PUBLIC_*` variables that are baked into the static build at **build time**.

Reviewer: Less LLM
```suggestion
This section lists what you must set. The backend reads its configuration from environment
variables. A full annotated template is in `backend/.env.example`. The frontend reads a few
`NEXT_PUBLIC_*` variables that are baked into the static build at **build time**.
```

### 5.1 Minimum backend variables for production

| Variable | Set it to |
| --- | --- |
| `ADMIN_PASSWORD_HASH` | The bcrypt hash from step 4.1 (single-quoted). |
| `JWT_SECRET` | The random value from step 4.2. |
| `ENVIRONMENT` | `production` (makes the app refuse to start if a secret is missing). |
| `CORS_ORIGINS` | The exact public origin of your frontend, for example `https://quiz.example.com`. Comma-separate multiple. |
| `FRONTEND_ORIGIN` | The same public frontend origin (used to build participant join links). No trailing slash. |
| `ADMIN_REFRESH_COOKIE_SECURE` | `true` (any HTTPS deployment). |
| `COGNIRO_DATA_DIR` | The directory where data is stored, for example `/var/lib/cogniro`. This is what you back up. |

If the frontend is served from a **different site** than the API (for example
`quiz.example.com` for the UI and `api.example.com` for the API), also set
`ADMIN_REFRESH_COOKIE_SAMESITE=none`. If they share the same site (recommended, via one
reverse proxy), leave it at the `lax` default.

### 5.2 Optional backend variables

> These have sensible defaults; override only if needed. See

Reviewer: ```suggestion
These have sensible defaults, override only if needed. See
```

[TECHNICAL-DOCUMENTATION.md](TECHNICAL-DOCUMENTATION.md) section 8 for the full table.

- `JWT_EXPIRE_MINUTES` (default 15), `ADMIN_REFRESH_EXPIRE_DAYS` (default 7)
- `RESULT_RETENTION_DAYS` (default 30), `PURGE_INTERVAL_SECONDS` (default 3600)
- `MEDIA_PUBLIC_PREFIX` (default `/media/quiz-assets`)
- `MAX_QUIZ_IMPORT_*` size limits
- `PLAY_RATE_LIMIT_*` (rate limiting; if you run behind a reverse proxy, set

>   `PLAY_RATE_LIMIT_TRUST_X_FORWARDED_FOR=true` so per-IP limits use the real client IP)

Reviewer: Not explicit enough

Create a file `backend.env` (do not commit it) with your real values. The name is arbitrary
because Docker just reads it and injects the values into the container. This is not the same
as the `backend/.env` used in the non-Docker setup (section 7.1): with Docker you do not need
a `.env` inside the image at all.

```
ADMIN_PASSWORD_HASH='$2b$12$your-real-hash-here'
JWT_SECRET=your-real-random-secret
ENVIRONMENT=production
CORS_ORIGINS=https://quiz.example.com
FRONTEND_ORIGIN=https://quiz.example.com
ADMIN_REFRESH_COOKIE_SECURE=true
COGNIRO_DATA_DIR=/var/lib/cogniro
```

Create a persistent directory on the host for the data and run the backend with the env file
and a mounted volume so data survives container restarts and rebuilds:

```bash
mkdir -p /srv/cogniro-data

docker run -d --name cogniro-backend \
  --env-file backend.env \
  -v /srv/cogniro-data:/var/lib/cogniro \
  -p 8000:8000 \
  --restart unless-stopped \
  cogniro-backend:latest
```

Run the frontend:

```bash
docker run -d --name cogniro-frontend \
  -p 3000:3000 \
  --restart unless-stopped \
  cogniro-frontend:latest
```

At this point the backend answers on port 8000 and the frontend on port 3000. In production

> you put both behind a reverse proxy with TLS (section 8). The volume mount

Reviewer: a comma is missing  
```suggestion
At this point the backend answers on port 8000 and the frontend on port 3000. In production,
you put both behind a reverse proxy with TLS (section 8). The volume mount
```

`-v /srv/cogniro-data:/var/lib/cogniro` is the single most important line: it is where all
quizzes and results are stored. Back it up (section 9).

### 6.4 A note on workers

Do not change the backend command to use multiple Uvicorn workers. The application keeps live
sessions and locks in process memory and is designed to run as a single process. Running
multiple workers will cause race conditions and split sessions. See
[TECHNICAL-DOCUMENTATION.md](TECHNICAL-DOCUMENTATION.md) section 6.5. To handle more load, use
a bigger host, not more workers.

---

## 7. Option B: Run without Docker

Here you install the runtimes directly on the host. You run the backend as a long-lived
service, build the frontend once into static files, and serve those files with a web server.

### 7.1 Backend: install and run

On the backend host:

```bash
# 1. Go to the backend directory
cd cogniro/backend

# 2. Install dependencies into a virtual environment (uv reads pyproject.toml + uv.lock)
uv sync --frozen --no-dev
```

**Step 3: create the `backend/.env` file.** This is how you configure the backend in the
non-Docker setup. The backend loads it automatically on startup (`main.py` calls

> `load_dotenv()`), so as long as the file sits in the `backend/` directory and you start the

Reviewer: missing comma
```suggestion
`load_dotenv()`), so as long as the file sits in the `backend/` directory, and you start the
```

Author: That comma is redundant, won't apply.
